cpp/src/io/parquet/predicate_pushdown.cpp

```diff
@@ -0,0 +1,527 @@
+    std::transform(per_file_metadata.begin(),
+                   per_file_metadata.end(),
+                   std::back_inserter(all_row_group_indices),
+                   [](auto const& file_meta) {
+                     std::vector<size_type> rg_idx(file_meta.row_groups.size());
+                     std::iota(rg_idx.begin(), rg_idx.end(), 0);
+                     return rg_idx;
+                   });
+    input_row_group_indices = host_span<std::vector<size_type> const>(all_row_group_indices);
+  } else {
+    input_row_group_indices = row_group_indices;
+  }
+  auto const total_row_groups = std::accumulate(input_row_group_indices.begin(),
+                                                input_row_group_indices.end(),
+                                                0,
+                                                [](size_type sum, auto const& per_file_row_groups) {
+                                                  return sum + per_file_row_groups.size();
+                                                });
+
+  // Converts Column chunk statistics to a table
+  // where min(col[i]) = columns[i*2], max(col[i])=columns[i*2+1]
+  // For each column, it contains #sources * #column_chunks_per_src rows.
+  std::vector<std::unique_ptr<column>> columns;
+  stats_caster stats_col{
+    total_row_groups, per_file_metadata, input_row_group_indices, cudf::get_default_stream()};
+  for (size_t col_idx = 0; col_idx < output_dtypes.size(); col_idx++) {
+    auto const& dtype = output_dtypes[col_idx];
+    // Only comparable types except fixed point are supported.
+    if (cudf::is_compound(dtype) && dtype.id() != cudf::type_id::STRING) {
+      // placeholder only for unsupported types.
+      columns.push_back(
+        cudf::make_numeric_column(data_type{cudf::type_id::BOOL8}, total_row_groups));
+      columns.push_back(
+        cudf::make_numeric_column(data_type{cudf::type_id::BOOL8}, total_row_groups));
+      continue;
+    }
+    auto [min_col, max_col] =
+      cudf::type_dispatcher<dispatch_storage_type>(dtype, stats_col, col_idx, dtype);
+    columns.push_back(std::move(min_col));
+    columns.push_back(std::move(max_col));
+  }
+  auto stats_table = cudf::table(std::move(columns));
+
+  // Converts AST to StatsAST with reference to min, max columns in above `stats_table`.
+  stats_expression_converter stats_expr{filter, static_cast<size_type>(output_dtypes.size())};
+  auto stats_ast = stats_expr.get_stats_expr();
+  auto predicate_col =
+    cudf::compute_column(stats_table, stats_ast.get(), rmm::mr::get_current_device_resource());
+  auto predicate = predicate_col->view();
+  CUDF_EXPECTS(predicate.type().id() == cudf::type_id::BOOL8,
+               "Filter expression must return a boolean column");
+
+  auto num_bitmasks = num_bitmask_words(predicate.size());
+  std::vector<bitmask_type> host_bitmask(num_bitmasks, ~bitmask_type{0});
+  if (predicate.nullable()) {
+    CUDF_CUDA_TRY(cudaMemcpy(host_bitmask.data(),
```

Comment:
Stream usage is not handled very well in this function. Currently we're shifting over all public APIs to accept streams (detail APIs should already take streams). I would like to establish the same conventions for the new APIs in this file.

Can we accept a stream in this function, and use it throughout? If not, can we at least declare a variable like `auto stream = cudf::get_default_stream()` and use it?

This line should use `cudaMemcpyAsync` and the rest of this function should use the same stream, where needed.

Reply:
cleaned up stream and mr usage.